qiskit/circuit/library/iqp.py

```diff
@@ -69,28 +76,79 @@ def __init__(self, interactions: list | np.ndarray) -> None:
-        a_str.replace("\n", ";")
-        name = "iqp:" + a_str.replace("\n", ";")
+def iqp(
+    interactions: Sequence[Sequence[int]] | None = None, num_qubits: int | None = None
+) -> QuantumCircuit:
+    r"""Instantaneous quantum polynomial time (IQP) circuit.
 
-        circuit = QuantumCircuit(num_qubits, name=name)
+    The circuit consists of a column of Hadamard gates, a column of powers of T gates,
+    a sequence of powers of CS gates (up to :math:`\frac{n^2-n}{2}` of them), and a final column of
+    Hadamard gates, as introduced in [1].
 
-        circuit.h(range(num_qubits))
-        for i in range(num_qubits):
-            for j in range(i + 1, num_qubits):
-                if interactions[i][j] % 4 != 0:
-                    circuit.cp(interactions[i][j] * np.pi / 2, i, j)
+    The circuit is parameterized by an :math:`n \times n` interactions matrix. The powers of each
+    T gate are given by the diagonal elements of the interactions matrix. The powers of the CS gates
+    are given by the upper triangle of the interactions matrix.
 
-        for i in range(num_qubits):
-            if interactions[i][i] % 8 != 0:
-                circuit.p(interactions[i][i] * np.pi / 8, i)
+    **Reference Circuit:**
 
-        circuit.h(range(num_qubits))
+    .. plot::
 
-        super().__init__(*circuit.qregs, name=circuit.name)
-        self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+       from qiskit.circuit.library import iqp
+       A = [[6, 5, 3], [5, 4, 5], [3, 5, 1]]
+       circuit = iqp(A)
+       circuit.draw("mpl")
+
+    **Expanded Circuit:**
+
+        .. plot::
+
+           from qiskit.circuit.library import iqp
+           from qiskit.visualization.library import _generate_circuit_library_visualization
+           A = [[6, 5, 3], [5, 4, 5], [3, 5, 1]]
+           circuit = iqp(A)
+           _generate_circuit_library_visualization(circuit)
+
+    **References:**
+
+    [1] M. J. Bremner et al. Average-case complexity versus approximate
+    simulation of commuting quantum computations,
+    Phys. Rev. Lett. 117, 080501 (2016).
+    `arXiv:1504.07999 <https://arxiv.org/abs/1504.07999>`_
+
+    Args:
+        interactions: The interactions as symmetric square matrix. If ``None``, then the
+            ``num_qubits`` argument must be set and a random IQP circuit will be generated.
+        num_qubits: If no interactions are given, construct a random IQP circuit with this
+            number of qubits. This argument cannot be passed at the same time as ``interactions``.
+
+    Returns:
+        An IQP circuit.
+    """
+    # if no interactions are given, generate them
+    if interactions is None:
+        if num_qubits is None:
+            raise ValueError("Either interactions or num_qubits must be provided.")
+
+    # otherwise validate the interactions
+    else:
+        if num_qubits is not None:
+            raise ValueError("Only one of interactions or num_qubits can be provided, not both.")
```

Comment:
perhaps both could be given, and one can assert that `num_qubits = len(interactions)` in this case ?

Reply:
Yeah I thought about that, it seemed a bit strange that passing an additional argument could raise an error in a function when one could just omit it. Explicitly disallowing passing both would make sure users can only have a single way of running the code... but I'm also happy to allow `num_qubits` 🙂